File: worker/src/fastcast.js
// fastcast.js — pure functions for ESPN's FastCast push layer (fastcast-plan.md).
// No I/O. Three exports, each with a faithful Dart port in
// app/lib/src/data/fastcast.dart (byte-for-byte parity via
// app/test/port_fastcast_test.dart against goldens from gen-goldens.mjs):
//
//   applyOps(doc, ops)       — standard RFC 6902 JSON Patch (gp-* topics).
//   applyEventOps(doc, ops)  — the event-* variant: paths are uid-prefixed
//                              ("s:1~l:10~e:401816076~c:401816076/situation/balls"),
//                              resolved to the event inside the doc, remainder
//                              applied as a standard pointer within it.
//   normalizeFastcastSlate(reg, key, doc) — event-* doc → per-event PARTIAL
//                              updates (score/phase/clock/detail/situation/
//                              seriesSummary), merged over the last polled
//                              canonical slate by the provider (Track 2). It is
//                              an overlay, NOT a scoreboard replacement — the
//                              event doc has no competitions[]/linescores/
//                              leaders, and (verified) no soccer goal timeline.
//
// Resilience rule (protocol): NEVER throw on a bad patch — apply what resolves,
// report what didn't. Both appliers return {doc, errors}; a non-empty errors
// array is the caller's signal to resync (refetch checkpoint / one poll).
// Both appliers deep-copy the input doc; the input is never mutated.

import { statusToPhase, buildScore, buildSituation } from './normalize.js';
import { resolve } from '../../schema/tools/resolve.mjs';

// JSON-safe deep copy (fixture/patch data is always JSON; keep the copy rule
// trivially portable to Dart — no structuredClone).
function deepCopy(v) {
  if (Array.isArray(v)) return v.map(deepCopy);
  if (v && typeof v === 'object') { const o = {}; for (const k of Object.keys(v)) o[k] = deepCopy(v[k]); return o; }
  return v;
}

// RFC 6901 pointer segment unescape (~1 → /, ~0 → ~; order matters).
const unescapeSeg = (s) => s.replace(/~1/g, '/').replace(/~0/g, '~');

// Split a root-relative pointer ("/a/b") into unescaped segments.
function segsOf(path) {
  if (path === '') return [];
  return path.split('/').slice(1).map(unescapeSeg);
}

// Walk to the PARENT of the pointer target. Returns {parent, key} or a string error.
function walk(root, segs) {
  let node = root;
  for (let i = 0; i < segs.length - 1; i++) {
    const seg = segs[i];
    if (Array.isArray(node)) {
      const idx = /^\d+$/.test(seg) ? +seg : -1;
      if (idx < 0 || idx >= node.length) return `no such index '${seg}'`;
      node = node[idx];
    } else if (node && typeof node === 'object') {
      if (!(seg in node)) return `no such key '${seg}'`;
      node = node[seg];
    } else return `not a container at '${seg}'`;
  }
  return { parent: node, key: segs[segs.length - 1] };
}

// Read the value at segs (for move/copy/test). Returns {value} or a string error.
function readAt(root, segs) {
  if (segs.length === 0) return { value: root };
  const w = walk(root, segs);
  if (typeof w === 'string') return w;
  const { parent, key } = w;
  if (Array.isArray(parent)) {
    const idx = /^\d+$/.test(key) ? +key : -1;
    if (idx < 0 || idx >= parent.length) return `no such index '${key}'`;
    return { value: parent[idx] };
  }
  if (parent && typeof parent === 'object') {
    if (!(key in parent)) return `no such key '${key}'`;
    return { value: parent[key] };
  }
  return `not a container at '${key}'`;
}

function deepEqual(a, b) {
  if (a === b) return true;
  if (Array.isArray(a) && Array.isArray(b)) return a.length === b.length && a.every((x, i) => deepEqual(x, b[i]));
  if (a && b && typeof a === 'object' && typeof b === 'object' && !Array.isArray(a) && !Array.isArray(b)) {
    const ka = Object.keys(a), kb = Object.keys(b);
    return ka.length === kb.length && ka.every((k) => deepEqual(a[k], b[k]));
  }
  return false;
}

// Apply ONE parsed op at segs within root. Returns null (ok) or a string error.
// `replace` is deliberately lenient (acts as set): FastCast replaces paths it
// never added, and strict-RFC existence checks would spray resyncs for nothing.
function applyAt(root, op, segs, value) {
  if (segs.length === 0) return 'root op unsupported'; // never observed; docs are replaced via checkpoint
  const w = walk(root, segs);
  if (typeof w === 'string') return w;
  const { parent, key } = w;
  if (Array.isArray(parent)) {
    if (op === 'add' && key === '-') { parent.push(value); return null; }
    const idx = /^\d+$/.test(key) ? +key : -1;
    if (idx < 0) return `bad array index '${key}'`;
    if (op === 'add') {
      if (idx > parent.length) return `index '${key}' out of range`;
      parent.splice(idx, 0, value); return null;
    }
    if (idx >= parent.length) return `no such index '${key}'`;
    if (op === 'remove') { parent.splice(idx, 1); return null; }
    if (op === 'replace') { parent[idx] = value; return null; }
    if (op === 'test') return deepEqual(parent[idx], value) ? null : 'test failed';
    return `unsupported op '${op}'`;
  }
  if (parent && typeof parent === 'object') {
    if (op === 'add' || op === 'replace') { parent[key] = value; return null; }
    if (op === 'remove') {
      if (!(key in parent)) return `no such key '${key}'`;
      delete parent[key]; return null;
    }
    if (op === 'test') return deepEqual(parent[key], value) ? null : 'test failed';
    return `unsupported op '${op}'`;
  }
  return `not a container at '${key}'`;
}

// Shared driver: `locate(path)` maps an op's path to {root, segs} (or a string
// error) — that's the only difference between the standard and event variants.
function applyWith(doc, ops, locate) {
  const out = deepCopy(doc);
  const errors = [];
  const list = Array.isArray(ops) ? ops : [];
  for (let i = 0; i < list.length; i++) {
    const o = list[i] || {};
    const op = o.op;
    const fail = (why) => errors.push(`${i}:${op} ${o.path}: ${why}`);
    if (typeof o.path !== 'string') { fail('no path'); continue; }
    const loc = locate(out, o.path);
    if (typeof loc === 'string') { fail(loc); continue; }
    if (op === 'add' || op === 'replace' || op === 'test') {
      const err = applyAt(loc.root, op, loc.segs, deepCopy(o.value));
      if (err) fail(err);
    } else if (op === 'remove') {
      const err = applyAt(loc.root, 'remove', loc.segs);
      if (err) fail(err);
    } else if (op === 'move' || op === 'copy') {
      if (typeof o.from !== 'string') { fail('no from'); continue; }
      const fromLoc = locate(out, o.from);
      if (typeof fromLoc === 'string') { fail(fromLoc); continue; }
      const r = readAt(fromLoc.root, fromLoc.segs);
      if (typeof r === 'string') { fail(r); continue; }
      const val = deepCopy(r.value);
      if (op === 'move') {
        const err = applyAt(fromLoc.root, 'remove', fromLoc.segs);
        if (err) { fail(err); continue; }
      }
      const err = applyAt(loc.root, 'add', loc.segs, val);
      if (err) fail(err);
    } else fail(`unsupported op '${op}'`);
  }
  return { doc: out, errors };
}

/** Standard RFC 6902 patch (gp-* topics use root-relative paths). */
export function applyOps(doc, ops) {
  return applyWith(doc, ops, (root, path) =>
    path === '' || path.startsWith('/') ? { root, segs: segsOf(path) } : `non-rooted path '${path}'`);
}

// Find an event by uid across doc.sports[].leagues[].events[].
function findEventByUid(doc, uid) {
  for (const sport of doc?.sports || []) {
    for (const lg of sport?.leagues || []) {
      for (const ev of lg?.events || []) if (ev?.uid === uid) return ev;
    }
  }
  return null;
}

/**
 * event-* topic patch: paths are "<uid>/<pointer...>" where <uid> is the raw
 * event uid (contains '~' and ':' — literal, NOT RFC 6901-escaped; split on '/'
 * BEFORE unescaping). A root-relative path ('/x' or '') applies standardly.
 */
export function applyEventOps(doc, ops) {
  return applyWith(doc, ops, (root, path) => {
    if (path === '' || path.startsWith('/')) return { root, segs: segsOf(path) };
    const slash = path.indexOf('/');
    const uid = slash === -1 ? path : path.slice(0, slash);
    const rest = slash === -1 ? '' : path.slice(slash);
    const ev = findEventByUid(root, uid);
    if (!ev) return `no event with uid '${uid}'`;
    if (rest === '') return `bare uid path`; // replacing a whole event needs a pointer
    return { root: ev, segs: segsOf(rest) };
  });
}

// ---- Track 2 normalizer -------------------------------------------------------
// The event-* doc, flattened per event (no competitions[]): status is a bare
// STRING ('in'), the real status object is `fullStatus` — so the house rule
// (branch on status.type.name, never state alone) reads fullStatus.type.
// Emits the canonical-aligned PARTIAL fields the overlay merge writes onto the
// polled competition; field names/derivations mirror buildCompetition exactly.
export function normalizeFastcastSlate(reg, key, doc) {
  const profile = resolve(reg, key);
  const wantId = profile.espnLeagueId != null ? String(profile.espnLeagueId) : null;
  const slug = key.split('/')[1];
  let league = null;
  for (const sport of doc?.sports || []) {
    for (const lg of sport?.leagues || []) {
      if (String(lg?.id ?? '') === wantId || lg?.slug === slug) { league = lg; break; }
    }
    if (league) break;
  }
  const events = [];
  for (const ev of league?.events || []) {
    const st = ev.fullStatus || {};
    const type = st.type || {};
    const ph = statusToPhase(type);
    const e = {
      id: String(ev.id ?? ''),
      status: {
        phase: ph.phase, live: ph.live, ended: ph.ended,
        period: typeof st.period === 'number' ? st.period : 0,
        periodLabel: type.shortDetail || type.detail || type.description || '',
        espnName: type.name || '', detail: type.detail || '',
      },
      competitors: (ev.competitors || []).map((c) => {
        const out = { id: String(c.id ?? '') };
        if (c.homeAway) out.homeAway = c.homeAway;
        if (c.score != null) out.score = buildScore(profile.scoreKind, c.score);
        if (typeof c.winner === 'boolean') out.winner = c.winner;
        return out;
      }),
    };
    if (ev.uid) e.uid = ev.uid;
    if (type.shortDetail) e.status.shortDetail = type.shortDetail;
    if (ph.live && st.displayClock && st.displayClock !== '0:00') e.status.clock = st.displayClock;
    // situation is scoreboard-shaped on the fastcast event (outsText rides the
    // event, not the situation) — reuse the scoreboard builder verbatim.
    const situation = buildSituation({ situation: ev.situation, outsText: ev.outsText });
    if (situation) e.situation = situation;
    if (ev.seriesSummary) e.seriesSummary = ev.seriesSummary;
    events.push(e);
  }
  return { key, events };
}

File: app/lib/src/data/fastcast.dart
// fastcast.dart — faithful Dart port of worker/src/fastcast.js (the FastCast
// pure layer: RFC 6902 application + the uid-prefixed event variant + the
// Track-2 slate overlay normalizer). Byte-for-byte parity with the JS oracle is
// pinned by test/port_fastcast_test.dart replaying the captured push streams
// (test/fixtures/golden/fastcast/). See fastcast-plan.md for the protocol.
//
// Resilience rule (protocol): NEVER throw on a bad patch — apply what resolves,
// report what didn't. Both appliers return {'doc':..., 'errors': [...]}; a
// non-empty errors list is the caller's signal to resync. Inputs are deep-copied,
// never mutated.

import 'profiles.dart';
import 'normalize.dart' show statusToPhase, buildScore, buildSituation;
import 'util.dart';

dynamic _deepCopy(dynamic v) {
  if (v is List) return v.map(_deepCopy).toList();
  if (v is Map) return {for (final k in v.keys) k: _deepCopy(v[k])};
  return v;
}

// RFC 6901 pointer segment unescape (~1 → /, ~0 → ~; order matters).
String _unescapeSeg(String s) => s.replaceAll('~1', '/').replaceAll('~0', '~');

List<String> _segsOf(String path) {
  if (path == '') return [];
  return path.split('/').skip(1).map(_unescapeSeg).toList();
}

final _digits = RegExp(r'^\d+$');

// Walk to the PARENT of the pointer target. Returns (parent, key) or an error string.
dynamic _walk(dynamic root, List<String> segs) {
  dynamic node = root;
  for (var i = 0; i < segs.length - 1; i++) {
    final seg = segs[i];
    if (node is List) {
      final idx = _digits.hasMatch(seg) ? int.parse(seg) : -1;
      if (idx < 0 || idx >= node.length) return "no such index '$seg'";
      node = node[idx];
    } else if (node is Map) {
      if (!node.containsKey(seg)) return "no such key '$seg'";
      node = node[seg];
    } else {
      return "not a container at '$seg'";
    }
  }
  return (node, segs[segs.length - 1]);
}

// Read the value at segs (for move/copy). Returns (value,) or an error string.
dynamic _readAt(dynamic root, List<String> segs) {
  if (segs.isEmpty) return (root,);
  final w = _walk(root, segs);
  if (w is String) return w;
  final (parent, key) = w as (dynamic, String);
  if (parent is List) {
    final idx = _digits.hasMatch(key) ? int.parse(key) : -1;
    if (idx < 0 || idx >= parent.length) return "no such index '$key'";
    return (parent[idx],);
  }
  if (parent is Map) {
    if (!parent.containsKey(key)) return "no such key '$key'";
    return (parent[key],);
  }
  return "not a container at '$key'";
}

bool _deepEqual(dynamic a, dynamic b) {
  if (identical(a, b)) return true;
  if (a is List && b is List) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (!_deepEqual(a[i], b[i])) return false;
    }
    return true;
  }
  if (a is Map && b is Map) {
    if (a.length != b.length) return false;
    for (final k in a.keys) {
      if (!b.containsKey(k) || !_deepEqual(a[k], b[k])) return false;
    }
    return true;
  }
  return a == b;
}

// Apply ONE op at segs within root. Returns null (ok) or an error string.
// `replace` is deliberately lenient (set semantics) — mirrors the JS oracle.
String? _applyAt(dynamic root, String op, List<String> segs, [dynamic value]) {
  if (segs.isEmpty) return 'root op unsupported';
  final w = _walk(root, segs);
  if (w is String) return w;
  final (parent, key) = w as (dynamic, String);
  if (parent is List) {
    if (op == 'add' && key == '-') {
      parent.add(value);
      return null;
    }
    final idx = _digits.hasMatch(key) ? int.parse(key) : -1;
    if (idx < 0) return "bad array index '$key'";
    if (op == 'add') {
      if (idx > parent.length) return "index '$key' out of range";
      parent.insert(idx, value);
      return null;
    }
    if (idx >= parent.length) return "no such index '$key'";
    if (op == 'remove') {
      parent.removeAt(idx);
      return null;
    }
    if (op == 'replace') {
      parent[idx] = value;
      return null;
    }
    if (op == 'test') return _deepEqual(parent[idx], value) ? null : 'test failed';
    return "unsupported op '$op'";
  }
  if (parent is Map) {
    if (op == 'add' || op == 'replace') {
      parent[key] = value;
      return null;
    }
    if (op == 'remove') {
      if (!parent.containsKey(key)) return "no such key '$key'";
      parent.remove(key);
      return null;
    }
    if (op == 'test') return _deepEqual(parent[key], value) ? null : 'test failed';
    return "unsupported op '$op'";
  }
  return "not a container at '$key'";
}

// Shared driver: `locate(root, path)` maps an op's path to (root, segs) or an
// error string — the only difference between the standard and event variants.
Map<String, dynamic> _applyWith(
    dynamic doc, dynamic ops, dynamic Function(dynamic root, String path) locate) {
  final out = _deepCopy(doc);
  final errors = <String>[];
  final list = ops is List ? ops : const [];
  for (var i = 0; i < list.length; i++) {
    final o = list[i] is Map ? list[i] as Map : const {};
    final op = o['op'];
    final path = o['path'];
    void failWith(String why) => errors.add('$i:$op $path: $why');
    if (path is! String) {
      failWith('no path');
      continue;
    }
    final loc = locate(out, path);
    if (loc is String) {
      failWith(loc);
      continue;
    }
    final (locRoot, locSegs) = loc as (dynamic, List<String>);
    if (op == 'add' || op == 'replace' || op == 'test') {
      final err = _applyAt(locRoot, op as String, locSegs, _deepCopy(o['value']));
      if (err != null) failWith(err);
    } else if (op == 'remove') {
      final err = _applyAt(locRoot, 'remove', locSegs);
      if (err != null) failWith(err);
    } else if (op == 'move' || op == 'copy') {
      final from = o['from'];
      if (from is! String) {
        failWith('no from');
        continue;
      }
      final fromLoc = locate(out, from);
      if (fromLoc is String) {
        failWith(fromLoc);
        continue;
      }
      final (fromRoot, fromSegs) = fromLoc as (dynamic, List<String>);
      final r = _readAt(fromRoot, fromSegs);
      if (r is String) {
        failWith(r);
        continue;
      }
      final val = _deepCopy((r as (dynamic,)).$1);
      if (op == 'move') {
        final err = _applyAt(fromRoot, 'remove', fromSegs);
        if (err != null) {
          failWith(err);
          continue;
        }
      }
      final err = _applyAt(locRoot, 'add', locSegs, val);
      if (err != null) failWith(err);
    } else {
      failWith("unsupported op '$op'");
    }
  }
  return {'doc': out, 'errors': errors};
}

/// Standard RFC 6902 patch (gp-* topics use root-relative paths).
/// Returns `{'doc': patched, 'errors': [String]}` — never throws.
Map<String, dynamic> applyOps(dynamic doc, dynamic ops) =>
    _applyWith(doc, ops, (root, path) => path == '' || path.startsWith('/')
        ? (root, _segsOf(path))
        : "non-rooted path '$path'");

dynamic _findEventByUid(dynamic doc, String uid) {
  for (final sport in (field(doc, 'sports') is List ? field(doc, 'sports') as List : const [])) {
    for (final lg in (field(sport, 'leagues') is List ? field(sport, 'leagues') as List : const [])) {
      for (final ev in (field(lg, 'events') is List ? field(lg, 'events') as List : const [])) {
        if (field(ev, 'uid') == uid) return ev;
      }
    }
  }
  return null;
}

/// event-* topic patch: paths are `<uid>/<pointer...>` where `<uid>` is the raw
/// event uid (contains '~' and ':' — literal, NOT RFC 6901-escaped; split on '/'
/// BEFORE unescaping). A root-relative path ('/x' or '') applies standardly.
Map<String, dynamic> applyEventOps(dynamic doc, dynamic ops) =>
    _applyWith(doc, ops, (root, path) {
      if (path == '' || path.startsWith('/')) return (root, _segsOf(path));
      final slash = path.indexOf('/');
      final uid = slash == -1 ? path : path.substring(0, slash);
      final rest = slash == -1 ? '' : path.substring(slash);
      final ev = _findEventByUid(root, uid);
      if (ev == null) return "no event with uid '$uid'";
      if (rest == '') return 'bare uid path';
      return (ev, _segsOf(rest));
    });

/// The event-* doc → per-event PARTIAL updates (Track 2 overlay): score, phase
/// (from `fullStatus.type` — the normal status object, house phase rules hold),
/// clock, detail, situation, seriesSummary. Merged over the last polled
/// canonical slate by the provider; never a scoreboard replacement.
Map<String, dynamic> normalizeFastcastSlate(Registry reg, String key, dynamic doc) {
  final profile = resolve(reg, key);
  final wantId = profile['espnLeagueId']?.toString();
  final slug = key.split('/')[1];
  Map? league;
  for (final sport in (field(doc, 'sports') is List ? field(doc, 'sports') as List : const [])) {
    for (final lg in (field(sport, 'leagues') is List ? field(sport, 'leagues') as List : const [])) {
      if (jsStr(field(lg, 'id') ?? '') == wantId || field(lg, 'slug') == slug) {
        league = lg as Map;
        break;
      }
    }
    if (league != null) break;
  }
  final events = <Map<String, dynamic>>[];
  for (final ev in (field(league, 'events') is List ? field(league, 'events') as List : const [])) {
    final st = ev['fullStatus'] is Map ? ev['fullStatus'] as Map : const {};
    final type = st['type'] is Map ? st['type'] as Map : const {};
    final ph = statusToPhase(type);
    final e = <String, dynamic>{
      'id': jsStr(ev['id'] ?? ''),
      'status': {
        'phase': ph['phase'], 'live': ph['live'], 'ended': ph['ended'],
        'period': st['period'] is num ? st['period'] : 0,
        'periodLabel': or([type['shortDetail'], type['detail'], type['description'], '']),
        'espnName': or([type['name'], '']), 'detail': or([type['detail'], '']),
      },
      'competitors': [
        for (final c in (ev['competitors'] is List ? ev['competitors'] as List : const []))
          (() {
            final out = <String, dynamic>{'id': jsStr(field(c, 'id') ?? '')};
            if (truthy(field(c, 'homeAway'))) out['homeAway'] = field(c, 'homeAway');
            if (field(c, 'score') != null) out['score'] = buildScore(profile['scoreKind'], field(c, 'score'));
            if (field(c, 'winner') is bool) out['winner'] = field(c, 'winner');
            return out;
          })(),
      ],
    };
    if (truthy(ev['uid'])) e['uid'] = ev['uid'];
    if (truthy(type['shortDetail'])) (e['status'] as Map)['shortDetail'] = type['shortDetail'];
    if (ph['live'] == true && truthy(st['displayClock']) && st['displayClock'] != '0:00') {
      (e['status'] as Map)['clock'] = st['displayClock'];
    }
    // situation is scoreboard-shaped on the fastcast event (outsText rides the
    // event, not the situation) — reuse the scoreboard builder verbatim.
    final situation = buildSituation({'situation': ev['situation'], 'outsText': ev['outsText']});
    if (situation != null) e['situation'] = situation;
    if (truthy(ev['seriesSummary'])) e['seriesSummary'] = ev['seriesSummary'];
    events.add(e);
  }
  return {'key': key, 'events': events};
}
